### ISSUE TITLE
Problems when using wrap_ad decorator

### ISSUE THREAD

**Issue by ligoudaner377** [mitsuba-renderer/mitsuba3#267]
## Summary

I'm trying to use pytorch and mitsuba at the same time. Similar to the tutorial [https://mitsuba.readthedocs.io/en/stable/src/inverse_rendering/pytorch_mitsuba_interoperability.html](url), I'm using the wrap_ad to convert torch tensor to drjit tensor. However, I got some errors.

## System configuration

System information:

  OS: Ubuntu 1804
  GPU: RTX3090
  Python version: 3.8.5
  CUDA version: 11.7
  NVidia driver: 515.65.01

  Dr.Jit version: 0.2.2
  Mitsuba version: 3.0.2
  Compiled with: pip installed

## Description

I'm trying to build a network to estimate the albedo value of BSDF and use mitsuba3 to train the network end to end. Here's a simplified code.
```python
import torch
import torch.nn as nn
import mitsuba as mi
import drjit as dr
mi.set_variant("cuda_ad_rgb")
scene = mi.load_file("../scenes/cbox.xml")
params = mi.traverse(scene)
key = 'red.reflectance.value'
ref_image = mi.render(scene, spp=256).torch()
@dr.wrap_ad(source='torch', target='drjit')
def render_albedo(albedo, spp=256, seed=1):
    # albedo: parameters that estimated by a neural network
    print(type(albedo))
    print(type(params[key]))
    params[key] = albedo
    params.update()
    return mi.render(scene, params, spp=spp, seed=seed, seed_grad=seed+1)

# define a tiny network
# input size: 256x256x3
# output size 3
class AlbedoModel(nn.Module):

    def __init__(self):
        super(AlbedoModel, self).__init__()
        self.layers = nn.Sequential(
            nn.Linear(3*256**2, 3),
            nn.Sigmoid(),
        )

    def forward(self, image):
        rgb = self.layers(image.view(-1))
        # Evaluate the model one channel as a time
        # Reconstruct and return the 3D tensor
        return rgb


model = AlbedoModel()
if 'cuda' in mi.variant():
    model = model.cuda()

optimizer = torch.optim.Adam(model.parameters(), lr=0.0002)
loss_fn = nn.L1Loss()
model.train(mode=True)
iteration_count = 10000
for i in range(iteration_count):
    optimizer.zero_grad()
    albedo_predicted = model(ref_image)
    image_predicted = render_albedo(albedo_predicted, seed=i, spp=8)
    loss = loss_fn(image_predicted, ref_image)
    loss.backward()
    optimizer.step()
```
However, I got errors like this
```python
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-f63f13ebce64> in <module>
     38     optimizer.zero_grad()
     39     albedo_predicted = model(ref_image)
---> 40     image_predicted = render_albedo(albedo_predicted, seed=i, spp=8)
     41     loss = loss_fn(image_predicted, ref_image)
     42     loss.backward()

~/anaconda3/lib/python3.8/site-packages/drjit/router.py in f(*args, **kwargs)
   5850                         return args_grad
   5851 
-> 5852                 return ToDrJit.apply(*args)
   5853 
   5854             if target == 'torch':

~/anaconda3/lib/python3.8/site-packages/drjit/router.py in forward(ctx, *args)
   5835                         ctx.args_drjit = torch_to_drjit(args)
   5836                         _dr.enable_grad(ctx.args_drjit)
-> 5837                         res = func(*ctx.args_drjit)
   5838                         ctx.res_drjit = (res,) if not isinstance(res, tuple) else res
   5839                         return drjit_to_torch(res)

<ipython-input-2-f63f13ebce64> in render_albedo(albedo, spp, seed)
      4     print(type(albedo))
      5     print(type(params[key]))
----> 6     params[key] = albedo
      7     params.update()
      8     return mi.render(scene, params, spp=spp, seed=seed, seed_grad=seed+1)

~/anaconda3/lib/python3.8/site-packages/mitsuba/python/util.py in __setitem__(self, key, value)
     79                     raise e
     80         else:
---> 81             self.set_property(cur, value_type, value)
     82 
     83     def __delitem__(self, key: str) -> None:

RuntimeError: Unable to cast Python instance to C++ type (#define PYBIND11_DETAILED_ERROR_MESSAGES or compile in debug mode for details)
``` 
I tried to print the data type of the estimated parameter albedo in wrap_ad function, and found that the torch tensor has been converted todrjit.cuda.ad.TensorXf
So I did a small experiment
```python
params[key] = mi.TensorXf([0.0, 0.0, 0.0], shape=[3])
``` 
and got the same error.
Is this an issue with the current wrap_ad function?

**Comment by Speierers** [comment]
Indeed the type of `params[key]` is `mi.Color3f` (a 3D static array) and unfortunately the current implementation of `mi.TensorXf` doesn't provide automatic conversion to this type, even when it has the right shape.

Since in you case you need to work with `mi.TensorXf` because of `dr.wrap_ad`, one solution would be to explicitly convert the tensor in the following way: 
```python
params[key] = dr.unravel(mi.Color3f, albedo)
```

Let me know if this work for you.

**Comment by ligoudaner377** [comment]
Thank you for your fast reply.
Unfortunately, after using dr.unravel, I got errors like this
```python
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-2-5f72b0c204cc> in <module>
     38     optimizer.zero_grad()
     39     albedo_predicted = model(ref_image)
---> 40     image_predicted = render_albedo(albedo_predicted, seed=i, spp=8)
     41     loss = loss_fn(image_predicted, ref_image)
     42     loss.backward()

~/anaconda3/lib/python3.8/site-packages/drjit/router.py in f(*args, **kwargs)
   5850                         return args_grad
   5851 
-> 5852                 return ToDrJit.apply(*args)
   5853 
   5854             if target == 'torch':

~/anaconda3/lib/python3.8/site-packages/drjit/router.py in forward(ctx, *args)
   5835                         ctx.args_drjit = torch_to_drjit(args)
   5836                         _dr.enable_grad(ctx.args_drjit)
-> 5837                         res = func(*ctx.args_drjit)
   5838                         ctx.res_drjit = (res,) if not isinstance(res, tuple) else res
   5839                         return drjit_to_torch(res)

<ipython-input-2-5f72b0c204cc> in render_albedo(albedo, spp, seed)
      4     print(type(albedo))
      5     print(type(params[key]))
----> 6     params[key] = dr.unravel(mi.Color3f, albedo)
      7     params.update()
      8     return mi.render(scene, params, spp=spp, seed=seed, seed_grad=seed+1)

~/anaconda3/lib/python3.8/site-packages/drjit/router.py in unravel(dtype, array, order)
   1120     if order == 'F':
   1121         indices = arange(index_type, n)
-> 1122         return gather(dtype, array, indices)
   1123     elif order == 'C':
   1124         result = dtype()

~/anaconda3/lib/python3.8/site-packages/drjit/router.py in gather(dtype, source, index, active)
    711     else:
    712         if _dr.is_tensor_v(dtype) or _dr.is_tensor_v(source):
--> 713             raise Exception("gather(): Tensor type not supported! Should work "
    714                             "with the underlying array instead. (e.g. tensor.array)")
    715         if source.Depth != 1:

Exception: gather(): Tensor type not supported! Should work with the underlying array instead. (e.g. tensor.array)
```

**Comment by Speierers** [comment]
Nice catch! This will be fixed in the next release: [578b0dd](https://github.com/mitsuba-renderer/drjit/commit/578b0dd6258995c95cd9a9213f1d7db39e93c0e9)

In the meantime you could simply do the following instead:
```python
params[key] = dr.unravel(mi.Color3f, albedo.array)
```
